toolbox/errorfinder/main.go

```diff
@@ -0,0 +1,509 @@
+	if err != nil {
+		log.Fatalf("Unable to retrieve data from sheet: %v", err)
+	}
+
+	if len(resp.Values) == 0 {
+		fmt.Println("No data found.")
+		return nil
+	}
+	filePaths := []string{}
+
+	for _, row := range resp.Values {
+		filePath := row[0]
+		filePathString, err := filePath.(string)
+		if err {
+			filePaths = append(filePaths, filePathString)
+		}
+	}
+
+	return filePaths
+}
+
+func NewErrorFinder(client *storage.Client, bucketName string) *ErrorFinder {
+	return &ErrorFinder{
+		client:     client,
+		bucketName: bucketName,
+	}
+}
+
+// Read from gcs the build-log.txt files into slice of lines.
+func (f *ErrorFinder) query(ctx context.Context, prefix string) ([]string, error) {
+	client := f.client
+	bucket := client.Bucket(f.bucketName)
+	buildFile := bucket.Object(prefix + "build-log.txt")
+
+	rc, err := buildFile.NewReader(ctx)
+	if err != nil {
+		return []string{}, err
+	}
+	defer rc.Close()
+	lines := []string{}
+
+	scanner := bufio.NewScanner(rc)
+	scanner.Split(bufio.ScanLines)
+	for scanner.Scan() {
+		line := scanner.Text()
+		if len(line) != 0 {
+			lines = append(lines, line)
+		}
+	}
+
+	return lines, nil
+}
+
+// Get error contents of files to update error map and warning map with new errors/warnings and path to build logs.
+func (f *ErrorFinder) getContent(
+	ctx context.Context, filePaths []string, errorMap map[string][]string, warningMap map[string][]string) (map[string][]string, map[string][]string) {
+	for _, filePath := range filePaths {
+		fileSlice, err := f.query(ctx, filePath)
+		if err != nil {
+			continue
+		}
+		output := []string{}
+		for _, line := range fileSlice {
+
+			// Ignore command lines to run that starts with "+".
+			if strings.Contains(line, "+") {
+				continue
+			}
+			// Generalize digits and content within quotations in current line.
+			generalized := f.generalizeDigits(line)
+			output = append(output, generalized)
+		}
+
+		// Loop through only generalized result lines.
+		for _, line := range output {
+			toLower := strings.ToLower(line)
+
+			// Only look at lines with key words of "error", "warning" and "fail".
+			if strings.Contains(toLower, "error") || strings.Contains(toLower, "failed") || strings.Contains(toLower, "failure") || strings.Contains(toLower, "warn") {
+
+				// Delete date information that are possibly present at the beginning of each line.
+				line = strings.TrimLeft(line, "\\d-T:.Z	info")
+				lineSplit := strings.Split(line, " ")
+				newLineSplit := []string{}
+
+				for indexD, section := range lineSplit {
+
+					// Ignore empty strings
+					if strings.Compare(section, "") == 0 {
+						continue
+					}
+
+					// If word does not contain key words, generalize word to characters necessary.
+					if !strings.Contains(strings.ToLower(section), "error") &&
+						!strings.Contains(strings.ToLower(section), "fail") && !strings.Contains(strings.ToLower(section), "warn") {
+						if indexD > 0 && strings.Compare(lineSplit[indexD-1], "container") == 0 {
```

Comment:
Can it be done with less ifs? Is it necessary to add \\s

Reply:
I added the \s-es because I think it would make the error message easier to understand. I'll try with fewer ifs